inet6.man
.SH INET6 ADDRESS FAMILY
This section documents the methods available in the
inet6 address family.
.SS The auto Method
This method may be used to define interfaces with automatically assigned IPv6 addresses. Using this method on its own doesn't mean that RDNSS options will be applied, too. To make this happen, \fBrdnssd\fP daemon must be installed, properly configured and running. If stateless DHCPv6 support is turned on, then additional network configuration parameters such as DNS and NTP servers will be retrieved from a DHCP server. Please note that on ifdown, the lease is not currently released (a known bug). 
.PP
.B Options
.RS
.TP
.BI privext " int           "
Privacy extensions (RFC4941) (0=off, 1=assign, 2=prefer)
.TP
.BI accept_ra " int         "
Accept router advertisements (0=off, 1=on, 2=on+forwarding). Default value: "2"
.TP
.BI dhcp " int              "
Use stateless DHCPv6 (0=off, 1=on)
.TP
.BI request_prefix " int    "
Request a prefix through DHCPv6 Prefix Delegation (0=off, 1=on). Default value: "0"
.TP
.BI ll-attempts
Number of attempts to wait for a link-local address. Default value: "60"
.TP
.BI ll-interval
Link-local address polling interval in seconds. Default value: "0.1"
.RE
.SS The loopback Method
This method may be used to define the IPv6 loopback interface. 
.PP
.B Options
.RS
.TP
(No options)
.RE
.SS The static Method
This method may be used to define interfaces with statically assigned IPv6 addresses. By default, stateless autoconfiguration is disabled for this interface. 
.PP
.B Options
.RS
.TP
.BI address " address       "
Address (colon delimited/netmask) \fBrequired\fP
.TP
.BI netmask " mask          "
Netmask (number of bits, eg 64)
.TP
.BI metric " metric         "
Routing metric for default gateway (integer)
.TP
.BI gateway " address       "
Default gateway (colon delimited)
.TP
.BI media " type            "
Medium type, driver dependent
.TP
.BI hwaddress " address     "
Hardware address or "random"
.TP
.BI mtu " size              "
MTU size
.TP
.BI accept_ra " int         "
Accept router advertisements (0=off, 1=on, 2=on+forwarding)
.TP
.BI autoconf " int          "
Perform stateless autoconfiguration (0=off, 1=on). Default value: "0"
.TP
.BI privext " int           "
Privacy extensions (RFC3041) (0=off, 1=assign, 2=prefer)
.TP
.BI scope
Address validity scope. Possible values: global, site, link, host
.TP
.BI preferred-lifetime " int"
Time that address remains preferred
.TP
.BI dad-attempts
Number of attempts to settle DAD (0 to disable DAD). Default value: "60"
.TP
.BI dad-interval
DAD state polling interval in seconds. Default value: "0.1"
.RE
.SS The manual Method
This method may be used to define interfaces for which no configuration is done by default. Such interfaces can be configured manually by means of \fBup\fP and \fBdown\fP commands or /etc/network/if-*.d scripts. 
.PP
.B Options
.RS
.TP
.BI hwaddress " address     "
Hardware address or "random"
.TP
.BI mtu " size              "
MTU size
.RE
.SS The dhcp Method
This method may be used to obtain network interface configuration via stateful DHCPv6 with dhclient. In stateful DHCPv6, the DHCP server is responsible for assigning addresses to clients. 
.PP
.B Options
.RS
.TP
.BI hwaddress " address     "
Hardware address or "random"
.TP
.BI accept_ra " int         "
Accept router advertisements (0=off, 1=on, 2=on+forwarding). Default value: "1"
.TP
.BI autoconf " int          "
Perform stateless autoconfiguration (0=off, 1=on)
.TP
.BI request_prefix " int    "
Request a prefix through DHCPv6 Prefix Delegation (0=off, 1=on). Default value: "0"
.TP
.BI ll-attempts
Number of attempts to wait for a link-local address. Default value: "60"
.TP
.BI ll-interval
Link-local address polling interval in seconds. Default value: "0.1"
.RE
.SS The v4tunnel Method
This method may be used to setup an IPv6-over-IPv4 tunnel. It requires the \fBip\fP command from the \fBiproute\fP package. 
.PP
.B Options
.RS
.TP
.BI address " address      "
Address (colon delimited) \fBrequired\fP
.TP
.BI netmask " mask         "
Netmask (number of bits, eg 64)
.TP
.BI endpoint " address     "
Address of other tunnel endpoint (IPv4 dotted quad) \fBrequired\fP
.TP
.BI local " address        "
Address of the local endpoint (IPv4 dotted quad)
.TP
.BI metric " metric        "
Routing metric for default gateway (integer)
.TP
.BI gateway " address      "
Default gateway (colon delimited)
.TP
.BI ttl " time             "
TTL setting
.TP
.BI mtu " size             "
MTU size
.TP
.BI preferred-lifetime " int"
Time that address remains preferred
.RE
.SS The 6to4 Method
This method may be used to setup an 6to4 tunnel. It requires the \fBip\fP command from the \fBiproute\fP package. 
.PP
.B Options
.RS
.TP
.BI local " address        "
Address of the local endpoint (IPv4 dotted quad) \fBrequired\fP
.TP
.BI metric " metric        "
Routing metric for default gateway (integer)
.TP
.BI ttl " time             "
TTL setting
.TP
.BI mtu " size             "
MTU size
.TP
.BI preferred-lifetime " int"
Time that address remains preferred
.RE


debian/changelog
ifupdown (0.8.19+2) unstable; urgency=medium

  * ifupdown-hotplug changes
  * networking to shut down interfaces at shutdown 

 -- anticapitalista <antix@operamail.com>  Tue, 03 Oct 2017 13:27:02 +0300

ifupdown (0.8.19+1) unstable; urgency=medium

  * antiX build 

 -- anticapitalista <antix@operamail.com>  Wed, 20 Sep 2017 00:24:45 +0100

ifupdown (0.8.19) unstable; urgency=medium

  * Fix exit code of ifquery --state. Closes: #852976

 -- Guus Sliepen <guus@debian.org>  Mon, 30 Jan 2017 14:32:18 +0100

ifupdown (0.8.18) unstable; urgency=medium

  [ Svante Signell ]
  * Makefile: Add removal of tests/*/state.* directories in the clean target.
  * Update testcases for GNU/Hurd.

  [ Guus Sliepen ]
  * Use dynamically allocated strings for filenames. This avoids using
    PATH_MAX, which caused a FTBFS on Hurd. Based on Svante's patch.
    Closes: #845140
  * Fix some memory leaks.
  * Improve error messages.
  * Always immediately quit when running out of memory.

 -- Guus Sliepen <guus@debian.org>  Wed, 11 Jan 2017 20:43:06 +0100

ifupdown (0.8.17) unstable; urgency=medium

  [ Scott Moser ]
  * Don't bring down the loopback interface when networking.service stops.
    Closes: #841106, #841107

  [ Guus Sliepen ]
  * Export $CLASS to scripts.
  * Add a bug-script.
  * Depend on freebsd-net-tools instead of net-tools.
  * Also configure VLANs on bridge interfaces (see #818849).
  * Improve manpage: simplify example, add more section headers.
    Closes: #840579
  * Provide IPv6 examples in /u/s/d/ifupdown/examples/network-interfaces.
    Closes: #703568
  * Use -4 option for IPv4 dhclient. Closes: #844592

  [ Martin Pitt ]
  * ifup@.service: start After the corresponding
    sys-subsystem-net-devices device.

 -- Guus Sliepen <guus@debian.org>  Tue, 10 Jan 2017 17:25:18 +0100

ifupdown (0.8.16) unstable; urgency=medium

  * Fix FTBFS due to testcases using the host's state information.

 -- Guus Sliepen <guus@debian.org>  Thu, 22 Sep 2016 13:20:18 +0200

ifupdown (0.8.15) unstable; urgency=medium

  * Fix ifdown complaining about vlan_id1. Closes: #838312
  * Add testcases for ifdown. 

 -- Guus Sliepen <guus@debian.org>  Thu, 22 Sep 2016 11:53:52 +0200

ifupdown (0.8.14) unstable; urgency=medium

  [ Benjamin Drung ]
  * Add InfiniBand parititon key support. Closes: #830212

  [ Richard Laager ]
  * Fix a crash when multiple interfaces are specified for no-scripts.
    Closes: #837732

  [ Bjørn Mork ]
  * Ignore link state when bringing up hotplug interfaces at boot.
    Closes: #814785, #834820

  [ Guus Sliepen ]
  * Add a test case for the no-scripts keyword.
  * Flush automatically assigned addresses on ifup when accept_ra=0.

 -- Guus Sliepen <guus@debian.org>  Sat, 17 Sep 2016 11:40:39 +0200

ifupdown (0.8.13) unstable; urgency=medium

  * Fix test suite on kFreeBSD. Closes: #826062
  * Check the return value of chdir() and ftruncate().
  * Clarify that scripts can be run with IFACE set to "--all".
    Closes: #826000

 -- Guus Sliepen <guus@debian.org>  Sat, 04 Jun 2016 14:08:54 +0200

ifupdown (0.8.12) unstable; urgency=medium

  * Add "nodad" to ip -6 addr add if dad-attempts is set to zero.
    Closes: #824686
  * Bump Standards-Version.

 -- Guus Sliepen <guus@debian.org>  Thu, 19 May 2016 15:36:02 +0200

ifupdown (0.8.11) unstable; urgency=medium

  [ Wido den Hollander ]
  * Wait properly for Link-Local Address to go through DAD.

  [ Imre Deak ]
  * Fix read_all_state when no state file exists. Closes: 819287

  [ Arthur Gautier ]
  * Implement link naming (ip link set alias)

  [ Martin Pitt <martin.pitt@ubuntu.com> ]
  * Add autopkgtest for "allow-hotplug" interfaces. Closes: #814312

  [ Guus Sliepen ]
  * Bump Standards-Version.

 -- Guus Sliepen <guus@debian.org>  Thu, 21 Apr 2016 18:22:20 +0200

ifupdown (0.8.10) unstable; urgency=medium

  * Allow "hwaddress random" to generate a random MAC address.
  * Allow ifdown/ifquery when physical interface is not same as logical
    interface. 

 -- Guus Sliepen <guus@debian.org>  Fri, 22 Jan 2016 16:35:23 +0100

ifupdown (0.8.9) unstable; urgency=medium

  * It's network-online.target, not .service.
  * Do not run "ip link set down" on manual interfaces during "ifdown -a".
    Closes: #809166
  * Add the keywords "no-auto-down" and "no-scripts". Closes: #615130

 -- Guus Sliepen <guus@debian.org>  Thu, 21 Jan 2016 22:02:11 +0100

ifupdown (0.8.8) unstable; urgency=medium

  [ Martin Pitt ]
  * Fix ifquery crash if interface state file does not exist yet.
    (Closes: #810779, LP: #1532722)
  * ifup@.service: Avoid stopping on shutdown via stopping system-ifup.slice
    (changed behaviour in systemd 228). (Closes: #761909, LP: #1492546)

  [ Guus Sliepen ]
  * Remove quotes around TimeoutStartSec parameter. Closes: #810656

 -- Guus Sliepen <guus@debian.org>  Tue, 12 Jan 2016 23:50:29 +0100

ifupdown (0.8.7) unstable; urgency=medium

  * Impose a 5 minute timeout for starting network services.
    Closes: #318577, #810656
  * Give an error when trying to up a dhcp interface when no DHCP client is
    installed. Closes: #471926, #419139
  * Add a section in the manual about extensions provided by other packages.
    Closes: #482405, #597274
  * Use absolute paths to binaries called by ifup/ifdown. Closes: #365114
  * Return an error when trying to ifup/ifdown/ifquery an unknown interface
    (unless it is ignored due to --allow or --exclude).
  * Ignore SIGPIPE. Closes: #311054
  * Update the bridge example. Closes: #488315
  * Use the onlink option when adding a gateway route. Closes: #378506
  * Also allow the metric option for tunnels and IPv6 interfaces.

 -- Guus Sliepen <guus@debian.org>  Mon, 11 Jan 2016 22:58:44 +0100

ifupdown (0.8.6) unstable; urgency=medium

  * Remove the Breaks: udev (<< 228-3~) to work around a bug in udev's prerm
    script. Closes: #809658, #809743
  * Run ip link up on manual interfaces, but ignore any errors.
  * Use sysctl to set the IPv6-specific MTU of an interface. Closes: #809714


inet6.c
#include <stddef.h>
#include "header.h"


#include "archcommon.h"
#include "archlinux.h"


static option_default _auto_default[] = {
	{ "accept_ra", "2" },
	{ "request_prefix", "0" },
	{ "ll-attempts", "60" },
	{ "ll-interval", "0.1" },
	{ NULL, NULL }
};
static conversion _auto_conv[] = {
	{ NULL, NULL, NULL, 0, NULL }
};
static int _auto_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/sbin/modprobe -q net-pf-10 > /dev/null 2>&1 || true # ignore failure.", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("-[[/sbin/sysctl -q -e -w net.ipv6.conf.%iface/.//%.use_tempaddr=%privext%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("-/sbin/sysctl -q -e -w net.ipv6.conf.%iface/.//%.accept_ra=%accept_ra%", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("-/sbin/sysctl -q -e -w net.ipv6.conf.%iface/.//%.autoconf=1", ifd, exec) && !ignore_failures) return 0;
}
if ( var_set("accept_ra", ifd) && !var_true("accept_ra", ifd) ) {
  if (!execute("/bin/ip addr flush dev %iface% mngtmpaddr", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip link set dev %iface% up", ifd, exec) && !ignore_failures) return 0;
}
if ( var_true("dhcp", ifd) && execable("/lib/ifupdown/wait-for-ll6.sh") ) {
  if (!execute("/lib/ifupdown/wait-for-ll6.sh", ifd, exec) && !ignore_failures) return 0;
}
if ( var_true("dhcp", ifd) && execable("/sbin/dhclient") && var_true("request_prefix", ifd) ) {
  if (!execute("/sbin/dhclient -6 -P -pf /run/dhclient6.%iface%.pid -lf /var/lib/dhcp/dhclient6.%iface%.leases -I -df /var/lib/dhcp/dhclient.%iface%.leases %iface%", ifd, exec) && !ignore_failures) return 0;
}
else if ( var_true("dhcp", ifd) && execable("/sbin/dhclient") ) {
  if (!execute("/sbin/dhclient -6 -S -pf /run/dhclient6.%iface%.pid -lf /var/lib/dhcp/dhclient6.%iface%.leases -I -df /var/lib/dhcp/dhclient.%iface%.leases %iface%", ifd, exec) && !ignore_failures) return 0;
}
else if ( var_true("dhcp", ifd) ) {
  if (!execute("echo 'No DHCPv6 client software found!' >/dev/stderr; false", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _auto_down(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/bin/ip -6 addr flush dev %iface% scope global", ifd, exec) && !ignore_failures) return 0;
}
if ( iface_is_link() ) {
  if (!execute("/bin/ip link set dev %iface% down", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _loopback_default[] = {
	{ NULL, NULL }
};
static conversion _loopback_conv[] = {
	{ NULL, NULL, NULL, 0, NULL }
};
static int _loopback_up(interface_defn *ifd, execfn *exec) {
if ( !iface_is_lo() ) {
  if (!execute("-/bin/ip link set dev %iface% up 2>/dev/null", ifd, exec) && !ignore_failures) return 0;
}
if ( !iface_is_lo() ) {
  if (!execute("-/bin/ip addr add dev %iface% ::1 2>/dev/null", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _loopback_down(interface_defn *ifd, execfn *exec) {
if ( !iface_is_lo() ) {
  if (!execute("-/bin/ip addr del dev %iface% ::1 2>/dev/null", ifd, exec) && !ignore_failures) return 0;
}
if ( !iface_is_lo() ) {
  if (!execute("-/bin/ip link set dev %iface% down 2>/dev/null", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _static_default[] = {
	{ "autoconf", "0" },
	{ "preferred-lifetime", "" },
	{ "dad-attempts", "60" },
	{ "dad-interval", "0.1" },
	{ NULL, NULL }
};
static conversion _static_conv[] = {
	{ "hwaddress", NULL, cleanup_hwaddress, 0, NULL },
	{ "preferred-lifetime", NULL, set_preferred_lft, 0, NULL },
	{ "address", "netmask?", get_token, 3, (char * []){"/", "1", ""} },
	{ "address", NULL, get_token, 3, (char * []){"/", "0", ""} },
	{ "gateway", "accept_ra?", if_set, 1, (char * []){"0"} },
	{ NULL, NULL, NULL, 0, NULL }
};
static int _static_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/sbin/modprobe -q net-pf-10 > /dev/null 2>&1 || true # ignore failure.", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("-[[/sbin/sysctl -q -e -w net.ipv6.conf.%iface/.//%.use_tempaddr=%privext%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("-[[/sbin/sysctl -q -e -w net.ipv6.conf.%iface/.//%.accept_ra=%accept_ra%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("-[[/sbin/sysctl -q -e -w net.ipv6.conf.%iface/.//%.autoconf=%autoconf%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("-[[/sbin/sysctl -q -e -w net.ipv6.conf.%iface/.//%.mtu=%mtu%]]", ifd, exec) && !ignore_failures) return 0;
}
if ( var_set("accept_ra", ifd) && !var_true("accept_ra", ifd) ) {
  if (!execute("/bin/ip addr flush dev %iface% mngtmpaddr", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip link set dev %iface% [[address %hwaddress%]] up", ifd, exec) && !ignore_failures) return 0;
}
if ( !var_set("dad-attempts", ifd) || var_true("dad-attempts", ifd) ) {
  if (!execute("/bin/ip -6 addr add %address%[[/%netmask%]] [[scope %scope%]] dev %iface% [[preferred_lft %preferred-lifetime%]]", ifd, exec) && !ignore_failures) return 0;
}
if ( var_set("dad-attempts", ifd) && !var_true("dad-attempts", ifd) ) {
  if (!execute("/bin/ip -6 addr add %address%[[/%netmask%]] [[scope %scope%]] dev %iface% [[preferred_lft %preferred-lifetime%]] nodad", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("[[ /bin/ip -6 route add default via %gateway% [[metric %metric%]] dev %iface% onlink ]]", ifd, exec) && !ignore_failures) return 0;
}
if ( !no_act && execable("/lib/ifupdown/settle-dad.sh") && (var_true("dad-attempts", ifd)) ) {
  if (!execute("/lib/ifupdown/settle-dad.sh", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _static_down(interface_defn *ifd, execfn *exec) {
{
  if (!execute("[[ /bin/ip -6 route del default via %gateway% [[metric %metric%]] dev %iface% ]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip -6 addr del %address%[[/%netmask%]] [[scope %scope%]] dev %iface%", ifd, exec) && !ignore_failures) return 0;
}
if ( iface_is_link() ) {
  if (!execute("/bin/ip link set dev %iface% down", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _manual_default[] = {
	{ NULL, NULL }
};
static conversion _manual_conv[] = {
	{ "hwaddress", NULL, cleanup_hwaddress, 0, NULL },
	{ NULL, NULL, NULL, 0, NULL }
};
static int _manual_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("-[[/sbin/sysctl -q -e -w net.ipv6.conf.%iface/.//%.mtu=%mtu%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("[[/bin/ip link set dev %iface% address %hwaddress%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip link set dev %iface% up 2>/dev/null || true", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _manual_down(interface_defn *ifd, execfn *exec) {
if ( iface_is_link() || !do_all ) {
  if (!execute("/bin/ip link set dev %iface% down 2>/dev/null || true", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _dhcp_default[] = {
	{ "accept_ra", "1" },
	{ "request_prefix", "0" },
	{ "ll-attempts", "60" },
	{ "ll-interval", "0.1" },
	{ NULL, NULL }
};
static conversion _dhcp_conv[] = {
	{ "hwaddress", NULL, cleanup_hwaddress, 0, NULL },
	{ NULL, NULL, NULL, 0, NULL }
};
static int _dhcp_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/sbin/modprobe -q net-pf-10 > /dev/null 2>&1 || true # ignore failure.", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("-[[/sbin/sysctl -q -e -w net.ipv6.conf.%iface/.//%.accept_ra=%accept_ra%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("-[[/sbin/sysctl -q -e -w net.ipv6.conf.%iface/.//%.autoconf=%autoconf%]]", ifd, exec) && !ignore_failures) return 0;
}
if ( var_set("accept_ra", ifd) && !var_true("accept_ra", ifd) ) {
  if (!execute("/bin/ip addr flush dev %iface% mngtmpaddr", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip link set dev %iface% [[address %hwaddress%]] up", ifd, exec) && !ignore_failures) return 0;
}
if ( execable("/lib/ifupdown/wait-for-ll6.sh") ) {
  if (!execute("/lib/ifupdown/wait-for-ll6.sh", ifd, exec) && !ignore_failures) return 0;
}
if ( execable("/sbin/dhclient") && var_true("request_prefix", ifd) ) {
  if (!execute("/sbin/dhclient -6 -pf /run/dhclient6.%iface%.pid -lf /var/lib/dhcp/dhclient6.%iface%.leases -I -P -N -df /var/lib/dhcp/dhclient.%iface%.leases %iface%", ifd, exec) && !ignore_failures) return 0;
}
else if ( execable("/sbin/dhclient") ) {
  if (!execute("/sbin/dhclient -6 -pf /run/dhclient6.%iface%.pid -lf /var/lib/dhcp/dhclient6.%iface%.leases -I -df /var/lib/dhcp/dhclient.%iface%.leases %iface%", ifd, exec) && !ignore_failures) return 0;
}
else if ( 1 ) {
  if (!execute("echo 'No DHCPv6 client software found!' >/dev/stderr; false", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _dhcp_down(interface_defn *ifd, execfn *exec) {
if ( execable("/sbin/dhclient") ) {
  if (!execute("/sbin/dhclient -6 -r -pf /run/dhclient6.%iface%.pid -lf /var/lib/dhcp/dhclient6.%iface%.leases -I -df /var/lib/dhcp/dhclient.%iface%.leases %iface%", ifd, exec) && !ignore_failures) return 0;
}
else if ( 1 ) {
  if (!execute("echo 'No DHCPv6 client software found!' >/dev/stderr; false", ifd, exec) && !ignore_failures) return 0;
}
if ( iface_is_link() ) {
  if (!execute("/bin/ip link set dev %iface% down", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _v4tunnel_default[] = {
	{ "preferred-lifetime", "" },
	{ NULL, NULL }
};
static conversion _v4tunnel_conv[] = {
	{ "preferred-lifetime", NULL, set_preferred_lft, 0, NULL },
	{ NULL, NULL, NULL, 0, NULL }
};
static int _v4tunnel_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/sbin/modprobe -q net-pf-10 > /dev/null 2>&1 || true # ignore failure.", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip tunnel add %iface% mode sit remote %endpoint% [[local %local%]] \
       [[ttl %ttl%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip link set %iface% up [[mtu %mtu%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("[[ /bin/ip addr add %address%[[/%netmask%]] dev %iface% [[preferred_lft %preferred-lifetime%]] ]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("[[ /bin/ip route add %gateway% dev %iface% ]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("[[ /bin/ip route add ::/0 via %gateway% [[metric %metric%]] dev %iface% onlink ]]", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _v4tunnel_down(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/bin/ip tunnel del %iface%", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _6to4_default[] = {
	{ "preferred-lifetime", "" },
	{ NULL, NULL }
};
static conversion _6to4_conv[] = {
	{ "local", "hexaddress", make_hex_address, 0, NULL },
	{ "preferred-lifetime", NULL, set_preferred_lft, 0, NULL },
	{ NULL, NULL, NULL, 0, NULL }
};
static int _6to4_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/sbin/modprobe -q net-pf-10 > /dev/null 2>&1 || true # ignore failure.", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip tunnel add %iface% mode sit remote any local %local% \
       [[ttl %ttl%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip link set %iface% up [[mtu %mtu%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip addr add 2002:%hexaddress%::1/16 dev %iface% [[preferred_lft %preferred-lifetime%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip route add 2000::/3 via ::192.88.99.1 [[metric %metric%]] dev %iface%", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _6to4_down(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/bin/ip -6 route flush dev %iface%", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip link set dev %iface% down", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip tunnel del %iface%", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static method methods[] = {
        {
                "6to4",
                _6to4_up, _6to4_down,
                _6to4_conv, _6to4_default
        },
        {
                "auto",
                _auto_up, _auto_down,
                _auto_conv, _auto_default
        },
        {
                "dhcp",
                _dhcp_up, _dhcp_down,
                _dhcp_conv, _dhcp_default
        },
        {
                "loopback",
                _loopback_up, _loopback_down,
                _loopback_conv, _loopback_default
        },
        {
                "manual",
                _manual_up, _manual_down,
                _manual_conv, _manual_default
        },
        {
                "static",
                _static_up, _static_down,
                _static_conv, _static_default
        },
        {
                "v4tunnel",
                _v4tunnel_up, _v4tunnel_down,
                _v4tunnel_conv, _v4tunnel_default
        },
};

address_family addr_inet6 = {
        "inet6",
        sizeof(methods)/sizeof(struct method),
        methods
};
